kotlin/workflow-core/src/main/java/com/squareup/workflow/WorkflowContext.kt

```diff
@@ -205,44 +209,61 @@ fun <StateT : Any, OutputT : Any, ChildStateT : Any, ChildRenderingT : Any>
 /**
  * Will wait for [deferred] to complete, then pass its value to [handler]. Once the handler has been
  * invoked for a given deferred+key, it will not be invoked again until an invocation of
- * [Workflow.compose] that does _not_ call this method with that deferred+key.
+ * [Workflow.compose] that does _not_ call this method with that deferred+[key].
+ *
+ * @param key A string key that is used to distinguish between subscriptions of the same type.
  */
 inline fun <reified T, StateT : Any, OutputT : Any> WorkflowContext<StateT, OutputT>.onDeferred(
   deferred: Deferred<T>,
   key: String = "",
   noinline handler: (T) -> WorkflowAction<StateT, OutputT>
 ) = onDeferred(
     deferred,
-    IdempotenceKey.fromGenericType(Deferred::class, T::class, key = key),
+    KTypes.fromGenericType(Deferred::class, T::class),
+    key,
     handler
 )
 
-@PublishedApi
-internal fun <T, StateT : Any, OutputT : Any> WorkflowContext<StateT, OutputT>.onDeferred(
+/**
+ * Will wait for [deferred] to complete, then pass its value to [handler]. Once the handler has been
+ * invoked for a given deferred+key, it will not be invoked again until an invocation of
+ * [Workflow.compose] that does _not_ call this method with that deferred+[key].
+ *
+ * @param type The [KType] that represents both the type of data source (e.g. `Deferred`) and the
+ * element type [T].
+ * @param key A string key that is used to distinguish between subscriptions of the same [type].
+ */
+fun <T, StateT : Any, OutputT : Any> WorkflowContext<StateT, OutputT>.onDeferred(
```

Comment:
You sure we want this to be public? Wondering if the ones that expose the `type` argument should all be `@PublishedApi` instead.

Reply:
It's needing to be public is the whole reason I started this PR. The inline functions can't be used by generic workflows.

Reply:
Right right right.